adit/dicom_web/utils/wadors_utils.py

```diff
@@ -96,3 +118,181 @@ def callback(ds: Dataset) -> None:
+        raise BadGatewayApiError(str(err))
+
+
+async def fetch_dicom_data(
+    source_server: DicomServer,
+    query: dict[str, str],
+    level: Literal["STUDY", "SERIES", "IMAGE"],
+) -> list[Dataset]:
+    """Fetch DICOM data using WadoFetcher and return the list of datasets"""
+    wado_fetcher = WadoFetcher(source_server, query)
+    dicom_images: list[Dataset] = []
+
+    def callback(ds: Dataset) -> None:
+        dicom_images.append(ds)
+
+    try:
+        await wado_fetcher.create_fetch_task(level, callback)
+
+    except RetriableDicomError as err:
+        raise ServiceUnavailableApiError(str(err))
+    except DicomError as err:
+        raise BadGatewayApiError(str(err))
+
+    return dicom_images
+
+
+async def process_single_fetch(dicom_images: list[Dataset]) -> AsyncIterator[tuple[str, BytesIO]]:
+    """
+    Process a list of DICOM datasets by converting them to NIfTI format
+    and yield the resulting files. Only handles conversion and yielding,
+    not the fetching of data.
+
+    For each file pair, yields the JSON file first, then the corresponding NIfTI file.
+
+    Only NoValidDicomError and NoSpatialDataError are silently ignored as these indicate
+    series that simply cannot be converted to NIfTI. All other exceptions are logged.
+    """
+    async with TemporaryDirectory() as temp_dir:
+        temp_path = Path(temp_dir)
+
+        # Write DICOM datasets to temporary files
+        for file_idx, dicom_image in enumerate(dicom_images):
+            dicom_file_path = temp_path / f"dicom_file_{file_idx}.dcm"
+            # Use sync_to_async to make write_dataset non-blocking
+            await sync_to_async(write_dataset, thread_sensitive=False)(dicom_image, dicom_file_path)
+
+        # Convert the DICOM files to NIfTI
+        nifti_output_dir = temp_path / "nifti_output"
+        await aiofiles.os.makedirs(nifti_output_dir, exist_ok=True)
+        converter = DicomToNiftiConverter()
+
+        try:
+            # Use sync_to_async for the converter
+            await sync_to_async(converter.convert, thread_sensitive=False)(
+                temp_path, nifti_output_dir
+            )
+        except (NoValidDicomError, NoSpatialDataError):
+            # These exceptions are expected for series that cannot be converted to NIfTI
+            # For example, non-image series (e.g., SR documents) or series without spatial data
+            # No warning needed, just skip this series without yielding any files
+            return
+        except DicomConversionError as e:
+            # Log warning for conversion errors
+            logger.warning(f"Failed to convert DICOM files to NIfTI: {e}")
+            # For conversion errors, there won't be any output files to process
+            return
+        except Exception as e:
+            # For all other unexpected errors, log and propagate the error
+            logger.error(f"Error during DICOM to NIfTI conversion: {e}")
+            raise
```

Comment:
I think I dislike this for the same reasons as mentioned in this comment: https://github.com/openradx/adit/pull/223#discussion_r2220205380
Only logging and re-raising doesn't make much sense. Not handled exceptions are logged anyway.